relay-filter/src/config.rs

```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+    fn is_version_supported(&self) -> bool {
+        if self.config.max_version < self.config.project_filters.version {
+            return false;
+        }
+        if let Some(gc) = self.config.global_filters {
+            if self.config.max_version < gc.version {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn next_project_filter(&mut self) -> Option<&'a GenericFilterConfig> {
+        let Some((id, filter)) = self
+            .config
+            .project_filters
+            .filters
+            .get_index(self.project_index)
+        else {
+            return None;
+        };
+        self.project_index += 1;
+
+        // Mark the filter as evaluated first. If the filter is disabled, it
+        // should not be applied from global configs.
+        let newly_evaluated = self.evaluated.insert(id);
+
+        if !filter.is_enabled || !newly_evaluated {
+            return None;
+        }
+
+        if filter.condition.is_none() {
+            let Some(global_filters) = self.config.global_filters else {
+                return None;
+            };
+            let Some(filter) = global_filters.filters.get(id) else {
+                return None;
+            };
+            return Some(filter);
+        }
+
+        Some(filter)
+    }
+
+    fn next_global_filter(&mut self) -> Option<&'a GenericFilterConfig> {
+        let Some(global_filters) = self.config.global_filters else {
+            return None;
+        };
+        let Some((id, filter)) = global_filters.filters.get_index(self.global_index) else {
+            return None;
+        };
+        self.global_index += 1;
+
+        if filter.is_empty() {
+            return None;
+        }
+
+        let newly_added = self.evaluated.insert(id);
+        if !newly_added {
+            return None;
+        }
+
+        Some(filter)
+    }
+}
+
+impl<'a> Iterator for DynamicFiltersConfigIter<'a> {
+    type Item = &'a GenericFilterConfig;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if !self.is_version_supported() {
+            return None;
+        }
+
+        // This loop starts by iterating through all project filters: increments
+        // `project_index` on every iteration until it grows too much (i.e. all
+        // project filters are evaluated), yielding the evaluated filter if
+        // necessary.
+        // Then, the same approach is applied to global filters.
+        // `continue` restarts the loop, so global filters aren't evaluated
+        // until all project filters are.
+        loop {
+            if self.project_index < self.config.project_filters.filters.len() {
+                let filter = self.next_project_filter();
+                if filter.is_some() {
+                    return filter;
+                }
+                continue;
+            }
+
+            if let Some(global_filters) = self.config.global_filters {
+                if self.global_index < global_filters.filters.len() {
```

Comment:
Might just be me but I find `if ... else if ... else` much easier to understand than multiple continue statements.

```suggestion
            } else if let Some(global_filters) = self.config.global_filters && self.global_index < global_filters.filters.len() {
```

Reply:
This seems to be an unstable feature, see https://github.com/rust-lang/rust/issues/53667. rust-analyzer also warns about the unnecessary `else` block.

Reply:
Followed a different and simpler approach in [a6db146](https://github.com/getsentry/relay/pull/3161/commits/a6db146fd2cbf1c653f989bbb05790e425e8023d).